### PR TITLE
[WIP] Building up the initial Chord and adding RPC layer

### PR DESCRIPTION
**Pull request by ulugbekna**

### CODE REVIEW

File: chord/helpers.go [hunk 1]
```diff
@@ -1,17 +1,39 @@
...
 }
+
+func hashString(s string) (string, error) {
+	sha256 := crypto.SHA256.New()
+	_, err := sha256.Write([]byte(s))
+	if err != nil {
+		return "", err
+	}
+	hash := sha256.Sum(nil)
+	return hex.EncodeToString(hash), nil
```
Comment (dosarudaniel) [inline chord/helpers.go]: We might need to keep only `m` bits

File: chord/chord.go [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
 	"google.golang.org/grpc"
-	"log"
+	"net"
+	"sync"
 )
 
-// IChordNode defines all of the functions that Chord will have (which I know at this moment)
+// IChordNode interface defines all of the functions that Chord will have (which I know at this moment)
 type IChordNode interface {
 	ChordServer
```
Comment (dosarudaniel) [inline chord/chord.go]: We need some comments that include the description of each function here

**Comment by ulugbekna** [comment]
Kept that for later.

**Comment by ulugbekna** [comment]
Feel free to add those from the paper

File: chord/chord.go [hunk 3]
```diff
@@ -32,9 +32,47 @@ type ChordNode struct {
...
+	chordNode.predecessor = nodeWithMux{
+		nil,
+		sync.RWMutex{},
+	}
+	chordNode.successorsList = successorsListWithMux{
+		make([]Node, 0),
+		sync.RWMutex{},
+	}
+	chordNode.fingerTable = fingerTableWithMux{
+		make(fingerTable, 0),
```
Comment (dosarudaniel) [inline chord/chord.go]: A finger table should have maximum `m` entries

**Comment by ulugbekna** [comment]
Could you add 'm' and 'r' cli args to main? m being the number of bits in a
node's id and r being the number of elements in the successor list?

**Comment by dosarudaniel** [comment]
> Could you add 'm' and 'r' cli args to main? m being the number of bits in a node's id and r being the number of elements in the successor list?

Sure!

File: chord/chord.go [hunk 4]
```diff
@@ -32,9 +32,47 @@ type ChordNode struct {
...
+		sync.RWMutex{},
+	}
+	chordNode.stubsPool = stubsPoolWithMux{
+		make(ipToStubMap),
+		sync.RWMutex{},
+	}
+
+	chordNode.chordServer = grpc.NewServer()
+	RegisterChordServer(chordNode.chordServer, &chordNode)
+	go chordNode.chordServer.Serve(listener)
```
Comment (dosarudaniel) [inline chord/chord.go]: There's an `unhandled error` _warning_ in GoLand for this.

File: chord/helpers.go (hunk 1 above)
Comment (ulugbekna) [reply chord/helpers.go]: > In my opinion we should close the opened connections after one RPC was executed - it's easy to implement this for the moment (We can analyze the impact of this).
`getStubFor` already works, so why bother

File: services/chord_service/chord_service.proto [hunk 5]
```diff
@@ -4,8 +4,9 @@ import "google/protobuf/empty.proto";
 package chord;
 
 service Chord {
-    rpc FindSuccessor (ID) returns (Nodes);
+    rpc FindSuccessor (ID) returns (Node);
```
Comment (dosarudaniel) [inline services/chord_service/chord_service.proto]: This function should return the first successor?

Comment (ulugbekna) [reply services/chord_service/chord_service.proto]: Yes

File: chord/chord.go [hunk 6]
```diff
@@ -110,10 +110,90 @@ func (chordNode *ChordNode) Create() {
...
+// Join lets the ChordNode to join a Chord ring containing node n0.
+// Algorithm:
+// n.join(n0)
+//   predecessor = nil;
+//   succ := n0.find_successor(n);
+//	 fingerTable[0] = succ
+//   succList := succ.getSuccessorsList();
+//   successorsList = succ :: succList[:len(succList) - 1]
+func (chordNode *ChordNode) Join(n0 Node) error {
+	var err error
```
Comment (dosarudaniel) [inline chord/chord.go]: Do you need this? Using `err :=` should do the job.

**Pull request merged**